libandroid/app/src/main/java/com/mapbox/services/android/testapp/distance/DistanceActivity.java

```diff
@@ -0,0 +1,340 @@
+import com.mapbox.mapboxsdk.style.sources.Source;
+import com.mapbox.services.android.testapp.R;
+import com.mapbox.services.android.testapp.Utils;
+import com.mapbox.services.commons.ServicesException;
+import com.mapbox.services.commons.geojson.Feature;
+import com.mapbox.services.commons.geojson.FeatureCollection;
+import com.mapbox.services.commons.geojson.Point;
+import com.mapbox.services.commons.models.Position;
+import com.mapbox.services.directions.v5.DirectionsCriteria;
+import com.mapbox.services.distance.v1.MapboxDistance;
+import com.mapbox.services.distance.v1.models.DistanceResponse;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+
+public class DistanceActivity extends AppCompatActivity implements OnMapReadyCallback, MapboxMap.OnMapClickListener {
+
+  private static final String TAG = "DistanceActivity";
+  private static final String KEY_DURATION = "duration";
+  private static final String KEY_NAME = "name";
+
+  private MapView mapView;
+  private MapboxMap mapboxMap;
+  private List<Position> coordinates;
+  private List<Feature> restaurants;
+
+  private List<Feature> features;
+
+  @Override
+  protected void onCreate(Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_distance);
+
+    Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
+    setSupportActionBar(toolbar);
+    getSupportActionBar().setDisplayHomeAsUpEnabled(true);
+
+    coordinates = new ArrayList<>();
+    restaurants = new ArrayList<>();
+
+    addRestaurants();
+
+    for (int i = 0; i < restaurants.size(); i++) {
+      coordinates.add((Position) restaurants.get(i).getGeometry().getCoordinates());
+    }
+
+    mapView = (MapView) findViewById(R.id.mapview);
+    mapView.onCreate(savedInstanceState);
+    mapView.getMapAsync(this);
+  }
+
+  @Override
+  public void onMapReady(MapboxMap mapboxMap) {
+    this.mapboxMap = mapboxMap;
+    mapboxMap.setOnMapClickListener(this);
+
+    for (int i = 0; i < restaurants.size(); i++) {
+      FeatureCollection featureCollection = FeatureCollection.fromFeatures(new Feature[] {restaurants.get(i)});
+      Source geoJsonSource = new GeoJsonSource(
+        restaurants.get(i).getStringProperty(KEY_NAME) + "-source", featureCollection
+      );
+      mapboxMap.addSource(geoJsonSource);
+    }
+
+    addMarkers();
+    updateLayers();
+  }
+
+  @Override
+  public void onMapClick(@NonNull LatLng point) {
+
+    features = new ArrayList<>();
+
+    final PointF pixel = mapboxMap.getProjection().toScreenLocation(point);
+    for (int i = 0; i < restaurants.size(); i++) {
+      features.addAll(mapboxMap.queryRenderedFeatures(
+        pixel,
+        restaurants.get(i).getStringProperty(KEY_NAME) + "-layer"
+      ));
+    }
+
+    features.addAll(mapboxMap.queryRenderedFeatures(pixel, "circle-layer"));
+
+    if (features.size() <= 0) {
```

Comment:
Might want to do the opposite here; `size() > 0` -> `callDistanceApi()` and have just a single exit point in the method